pkg/apis/configuration/v1alpha1/kongcustomentity_types.go

```diff
@@ -0,0 +1,144 @@
+// +kubebuilder:resource:path=kongcustomentities,shortName=kce,categories=kong-ingress-controller
+// +kubebuilder:subresource:status
+
+// KongCustomEntity represents a custom entity in Kong.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec is the specification of the entity.
+	Spec KongCustomEntitySpec `json:"spec"`
+	// Status is the status of the entity.
+	Status KongCustomEntityStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:
+
+// KongCustomEntitySpec defines the specification of a Kong custom entity.
+type KongCustomEntitySpec struct {
+	// +kubebuilder:validation:Required
+	// Type is the type of this custom entity.
+	// Should be same as the `Name` of a KongCustomEntityDefinition.
+	Type string `json:"type"`
+	// +kubebuilder:validation:Type=object
+	// Fields is the fields of the custom entity, in JSON format.
+	Fields apiextensionsv1.JSON `json:"fields"`
+	// Patches stores fields coming from external resources (e.g. secrets) adding to
+	// to a certain JSON path of the entity.
+	Patches []ConfigSourcePatch `json:"patches,omitempty"`
+}
+
+// +k8s:deepcopy-gen:
+
+// ConfigSourcePatch is a patch to add values from external resources (e.g. secrets)
+// to a certain JSON path.
+type ConfigSourcePatch struct {
+	// +kubebuilder:validation:Required
+	// Path is the JSON path of imported configurations to add to.
+	Path string `json:"path"`
+	// ConfigSource is the source secret name and key of the value.
+	ConfigSource *kongv1.ConfigSource `json:"configSource,omitempty"`
+}
+
+// +k8s:deepcopy-gen:
+
+// KongCustomEntityStatus defines the status of a Kong custom entity.
+type KongCustomEntityStatus struct {
+	// Conditions describe the current conditions of the KongCustomEntityStatus.
+	//
+	// Known condition types are:
+	//
+	// * "Programmed"
+	//
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,path=kongcustomentitydefinitions,shortName=kced,categories=kong-ingress-controller
+
+// KongCustomEntityDefinition represents definition of a custom entity type in Kong.
+type KongCustomEntityDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec KongCustomEntityDefinitionSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:
+
+// KongCustomEntityDefinitionSpec is the specification of KongCustomEntityDefinition
+// to define a type of Kong custom entity.
+type KongCustomEntityDefinitionSpec struct {
+	// +kubebuilder:validation:Required
+	// Name is the type name of the entity.
```

Comment:
```suggestion
	// Name is the type name of the entity.
	// +kubebuilder:validation:Required
```
This is the order that we have in other files